src/runtime_src/core/pcie/driver/linux/xocl/subdev/pcie_firewall.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2020 Xilinx, Inc. All rights reserved.
+ *
+ * Authors: Lizhi.Hou@xilinx.com
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/*
+ * the PCIe Firewall IP to protect against host access to BARs which are not
+ * available i.e. when the PLP is in reset, not yet configured or not
+ * implemented.
+ *
+ * Details of the IP are available at:
+ *     https://confluence.xilinx.com/display/XIP/PCIe+Firewall+IP+Development
+ *
+ * Following server warm/cold boot or hot reset, the PCIe Firewall will
+ * automatically respond to accesses to BARs implemented in the PLP for compute
+ * platforms i.e.
+ *    PF0, BAR2
+ *    PF1, BAR2
+ *    PF1, BAR4
+ * Once the PLP has been programmed and ep_pr_isolate_plp_00 has been released
+ * from reset, then XRT should program the PCIe Firewall IP to clear the
+ * appropriate bits in the Enable Response Register (0x8) to allow transactions
+ * to propagate to the PLP.
+ */
+
+#include "../xocl_drv.h"
+
+struct firewall_regs {
+	u32		fwr_ip_ver;
+	u32		fwr_cap;
+	u32		fwr_en_resp;
+	u32		fwr_pf0_resp_addr;
+	u32		fwr_pf1_resp_addr;
+	u32		fwr_pf2_resp_addr;
+	u32		fwr_pf3_resp_addr;
+} __attribute__((packed));
+
+struct firewall {
+	struct platform_device	*pdev;
+	void		__iomem	*base;
+	struct mutex		fw_lock;
+};
+
+#define reg_rd(g, r)					\
+	XOCL_READ_REG32(&((struct firewall_regs *)g->base)->r)
+#define reg_wr(g, v, r)					\
+	XOCL_WRITE_REG32(v, &((struct firewall_regs *)g->base)->r)
+
+#define UNBLOCK_BIT(pf, bar)					\
+	(1U << ((pf) * 6 + (bar)))
+
+static int firewall_unblock(struct platform_device *pdev, int pf, int bar)
+{
+	struct firewall *firewall = platform_get_drvdata(pdev);
+	u32 val;
+
+	mutex_lock(&firewall->fw_lock);
+	val = reg_rd(firewall, fwr_en_resp);
+	if (val & UNBLOCK_BIT(pf, bar)) {
+		xocl_info(&pdev->dev, "unblock pf%d, bar%d", pf, bar);
+		reg_wr(firewall, (~UNBLOCK_BIT(pf, bar)) & val, fwr_en_resp);
+	}
+	mutex_unlock(&firewall->fw_lock);
+
+	return 0;
+}
```

Comment:
Can the IP provide a single bit to unblock everything? When do we ever want to unblock one BAR but leave another blocked?